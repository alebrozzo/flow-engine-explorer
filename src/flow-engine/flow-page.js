// @flow
import * as React from "react";

import type { FlowFunction, ExecutionResult } from "../types";
import { FunctionViewer } from "./function-viewer";
import "./flow-page.css";

type FunctionResult = {
    fn: FlowFunction,
    passed: boolean,
};

type Props = {
    functions: FlowFunction[],
    executionOrder: ExecutionResult[],
};

const FlowPage = (props: Props) => {
    const functionsToDisplay: FunctionResult[] = [];
    for (const result of props.executionOrder) {
        // $FlowFixMe
        functionsToDisplay.push({ fn: props.functions.find(fn => fn.id === result.id), passed: result.result });
    }

    return (
        <div className="flow_page__container">
            <div className="flow_page__header">
                <h1>Flow Engine</h1>
            </div>
            <div className="flow_page__function_list">
                {functionsToDisplay.map(func => (
                    <FunctionViewer key={func.fn.id} flowFunction={func.fn} passed={func.passed} />
                ))}
            </div>
        </div>
    );
};
export { FlowPage };
