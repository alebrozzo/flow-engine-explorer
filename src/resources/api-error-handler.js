// @flow
export function normalizeApiErrors(err: any, logToConsole?: boolean): Array<string> {
    if (logToConsole) {
        console.error("API error:", err);
    }
    let errors: Array<string> = [];
    const statusCode = (err.response && err.response.status) || err.status || -1;
    if (statusCode === 500) {
        errors.push("There was an error accessing the server.");
    } else if (statusCode === 401 || statusCode === 403) {
        errors.push(
            "You do not have permissions to access this feature.\nPlease contact an Administrator for assistance."
        );
    } else if (statusCode === 404) {
        errors.push("The resource was not found.");
    } else if (statusCode === 400 && err.response && err.response.data) {
        Object.keys(err.response.data).forEach(key => {
            errors = errors.concat(err.response.data[key]);
        });
    } else {
        errors.push(err.message || "Unknown error.");
    }
    if (logToConsole) {
        console.error("API error list:", errors);
    }
    return errors;
}
