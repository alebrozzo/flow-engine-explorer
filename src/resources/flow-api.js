export const getFlowFunctions = () =>
    // simulate an API hit delay of a random time between 0.2 and 1 seconds
    new Promise((resolve, reject) => {
        setTimeout(() => resolve(getData()), Math.floor(Math.random() * 800) + 200);
    });

export const getData = () => [
    {
        id: 1,
        title: "Verify if weekend",
        body: `function isWeekend(date) {
  const dayOfWeek = date.getDay();
  return (dayOfWeek === 6) || (dayOfWeek === 0);
}`,
        true_id: 2,
        false_id: 3,
    },
    {
        id: 2,
        title: "Celebrate",
        body: `function celebrate() {
  console.log('Party on!!!');
  return true;
}`,
        true_id: null,
        false_id: null,
    },
    {
        id: 3,
        title: "Verify if Friday",
        body: `function isItFriday(date) {
  const dayOfWeek = date.getDay();
  return dayOfWeek === 5;
}`,
        true_id: 4,
        false_id: 5,
    },
    {
        id: 4,
        title: "TGIF",
        body: `function tgif() {
  console.log('TGIF!');
  return true;
}`,
        true_id: null,
        false_id: null,
    },
    {
        id: 5,
        title: "Groan",
        body: `function groan(date) {
  const daysUntilWeekend = 6 - date.getDay();
  console.log('Hang in there... ' + daysUntilWeekend + ' days until the weekend!');
  return true;
}`,
        true_id: null,
        false_id: null,
    },
];
