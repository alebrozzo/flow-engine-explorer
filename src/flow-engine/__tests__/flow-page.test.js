import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { getData } from "../../resources/flow-api";
import { FlowPage } from "../flow-page";

describe("FlowPage", () => {
    let defaults;
    beforeAll(() => {
        defaults = {
            functions: getData(),
            executionOrder: [{ id: 1, result: true }, { id: 2, result: false }],
        };
    });

    it("renders the page", () => {
        const component = shallow(<FlowPage functions={defaults.functions} executionOrder={defaults.executionOrder} />);
        expect(toJson(component)).toMatchSnapshot();
    });
});
