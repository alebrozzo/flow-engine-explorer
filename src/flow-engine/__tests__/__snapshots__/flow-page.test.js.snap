// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowPage renders the page 1`] = `
<div
  className="flow_page__container"
>
  <div
    className="flow_page__header"
  >
    <h1>
      Flow Engine
    </h1>
  </div>
  <div
    className="flow_page__function_list"
  >
    <FunctionViewer
      flowFunction={
        Object {
          "body": "function isWeekend(date = new Date()) {
          const dayOfWeek = date.getDay();
          return (dayOfWeek === 6) || (dayOfWeek === 0);
        }",
          "false_id": 5,
          "id": 1,
          "title": "Verify if weekend",
          "true_id": 2,
        }
      }
      key="1"
      passed={true}
    />
    <FunctionViewer
      flowFunction={
        Object {
          "body": "function celebrate() {
          console.log('Party on!!!');
          return true;
        }",
          "false_id": null,
          "id": 2,
          "title": "Celebrate",
          "true_id": null,
        }
      }
      key="2"
      passed={false}
    />
  </div>
</div>
`;
